### ISSUE TITLE
Introduce logging

### ISSUE THREAD

**Issue by ntnn** [romainl/vim-qf#22]
Introduce logging

Soft-touched logging would be nice, e.g. ``qf#log``, ``qf#debug`` and
``qf#error``, which prefix the message with a ``vim-qf:level:`` and the
debug level only showing when e.g. ``g:qf_debug`` is true or the like.

This'd help greatly with debugging in the future as well as help
understand the code (e.g. on early returns etc.pp. a debug message would
act like a comment explaining the early exit.

**Comment by arecarn** [comment]
For some of of my scripts I've used [Decho](https://github.com/vim-scripts/Decho). It is definitely not perfect but it is pretty easy to start using.

**Comment by ntnn** [comment]
@romainl what do you think? Do you want external dependencies?
Another possibility would be to write our own logging library with
structured logging - though that might be a bit oversized for a bit of
scripting.

Sent with AquaMail for Android
http://www.aqua-mail.com

**Comment by romainl** [comment]
Silly me… that was #23.

**Comment by romainl** [comment]
@ntnn I don't want external dependencies if I can avoid them. I'd like to play with `:throw` and friends before turning to a logging lib.

**Comment by arecarn** [comment]
A real vimscript logging library would be very nice, I could never seem find one.

Once potentially nice thing about Decho is that it is only a dependency for anyone who wants to use logging. In my case I just enabled it for debugging and disabled it before committing changes or fixes.

**Comment by ntnn** [comment]
No, having a difference in the code between dev and production is not a choice.
I'd rather use a stub file which defines the functions as noops if the logging plugin is not loaded.

@romainl Mhm... I don't like exceptions, I'd rather sanitize input more. 
A properly written logging library would be nice, on the other hand simple logging would be just a case of using a ``s:log`` function and ``qf#error``, ``qf#debug`` and ``qf#info``, wrapping the ``s:log``. Those three levels should be sufficient for most use.
